addons/costume-editor-shortcuts/userscript.js

```diff
@@ -0,0 +1,141 @@
+  addon.tab.redux.addEventListener("statechanged", handleStateChanged);
+
+  /**
+   * If costume editor is open, initialize tool shortcuts, otherwise clean them up if needed.
+   */
+  async function handleStateChanged(event) {
+    // If "Convert to Bitmap/Vector" button is pressed in the costume editor, re-draw shortcuts on the buttons.
+    if (event.detail.action.type === "scratch-paint/formats/CHANGE_FORMAT") {
+      setTimeout(async () => await addShortcutsToTitles(), 0); // allow the DOM to update before calling addLettersToButtons.
+      return;
+    }
+
+    // Only initialize or cleanup when the user switches to a new tab.
+    const activeIndex = addon.tab.redux.state.scratchGui.editorTab.activeTabIndex;
+    if (prevEditorTabIndex != activeIndex) {
+      prevEditorTabIndex = activeIndex;
+      if (activeIndex === COSTUME_EDITOR_TAB_INDEX) {
+        await initialize();
+      } else if (activeIndex !== COSTUME_EDITOR_TAB_INDEX && isInitialized) {
+        await cleanup();
+      }
+    }
+  }
+
+  /**
+   * Setup keydown listeners for shortcuts and and update tooltips to include shortcuts.
+   */
+  async function initialize() {
+    if (isInitialized) return; // Prevents double initialization
+    isInitialized = true;
+    document.addEventListener("keydown", handleKeyDown);
+    document.addEventListener("focusin", userStartedTyping);
+    document.addEventListener("focusout", userStoppedTyping);
+    await addShortcutsToTitles();
+  }
+
+  /**
+   * Remove keydown listeners.
+   */
+  async function cleanup() {
+    isInitialized = false;
+    document.removeEventListener("keydown", handleKeyDown);
+    document.removeEventListener("focusin", userStartedTyping);
+    document.removeEventListener("focusout", userStoppedTyping);
+  }
+
+  /**
+   * Switch costume editor tool if a valid shortcut was pressed.
+   */
+  async function handleKeyDown(event) {
+    if (isUserTyping) return;
+
+    var toolName = shortcutToToolName[event.key.toLowerCase()];
+    await switchTool(toolName);
+  }
+
+  /**
+   * For selecting costume editor tools by name.
+   */
+  async function switchTool(toolName) {
+    if (!toolName || addon.tab.redux.state.scratchGui.editorTab.activeTabIndex !== COSTUME_EDITOR_TAB_INDEX) return;
+
+    try {
+      var modeSelector = await addon.tab.waitForElement("[class^='paint-editor_mode-selector']");
+      if (modeSelector) {
+        modeSelector.querySelector(`span[title^='${toolName}']`)?.click();
+      }
+    } catch (error) {
+      console.error(`Failed to switch paint tool to: '${toolName}'. `, error);
+    }
+  }
+
+  /**
+   * Iterate over the costume editor tool buttons and update their tooltips to include shortcuts.
+   */
+  async function addShortcutsToTitles() {
+    try {
+      const container = await addon.tab.waitForElement("[class^='paint-editor_mode-selector']");
+      container.querySelectorAll("span").forEach((span) => {
+        updateTitle(span, span.getAttribute("title"));
```

Comment:
This will only work when the project editor is set to English.

@misaunde I'm not sure of any good solutions to identify which button is which, though.

Reply:
I almost forget we have a dedicated method for this.

Try comparing against `addon.tab.scratchMessage("paint.brushMode.brush")` instead of hardcoding the English string `"Brush"`.

To find the message ID for a given string, type `__addon.tab.redux.state.locales.messages` into the console and it will display all key/value pairs.

Reply:
The index varies on bitmap vs vector I think, so it's probably easier to compare the title attribute.